File: ios/HoloNetTests/Module/Forum/Data/ForumParserTests.swift
//
//  ForumParserTests.swift
//  SWTOR HoloNet
//
//  Created by Ivan Fabijanovic on 05/11/14.
//  Copyright (c) 2014 Ivan Fabijanović. All rights reserved.
//

import UIKit
import XCTest
import HTMLReader

class ForumParserTests: XCTestCase {
    fileprivate var parser: ForumParser!
    fileprivate var language: ForumLanguage!
    
    override func setUp() {
        super.setUp()
        
        self.parser = ForumParser()
        self.language = .english
    }
}

extension ForumParserTests {
    func testLinkParameter_Success() {
        let link = HTMLElement(tagName: "a", attributes: ["href": "http://www.holonet.test?param=value"])
        let param = self.parser!.linkParameter(linkElement: link, name: "param")!
        
        XCTAssertEqual(param, "value", "")
    }
    
    func testLinkParameter_ElementNil() {
        let param = self.parser!.linkParameter(linkElement: nil, name: "param")
        
        XCTAssertNil(param, "")
    }
    
    func testLinkParameter_NoParameters() {
        let link = HTMLElement(tagName: "a", attributes: ["href": "http://www.holonet.test"])
        let param = self.parser!.linkParameter(linkElement: link, name: "param")
        
        XCTAssertNil(param, "")
    }
    
    func testLinkParameter_MissingParameter() {
        let link = HTMLElement(tagName: "a", attributes: ["href": "http://www.holonet.test?param=value"])
        let param = self.parser!.linkParameter(linkElement: link, name: "otherParam")
        
        XCTAssertNil(param, "")
    }
    
    func testLinkParameter_MultipleParameters() {
        let link = HTMLElement(tagName: "a", attributes: ["href": "http://www.holonet.test?param1=value1&param2=value2"])
        let param = self.parser!.linkParameter(linkElement: link, name: "param1")!
        
        XCTAssertEqual(param, "value1", "")
    }
}

extension ForumParserTests {
    func testIntegerContent_Success() {
        let html = HTMLElement(tagName: "div", attributes: nil)
        html.textContent = "123"
        let value = self.parser!.integerContent(element: html)!
        
        XCTAssertEqual(value, 123, "")
    }
    
    func testIntegerContent_Nested() {
        let parent = HTMLElement(tagName: "div", attributes: nil)
        let child = HTMLElement(tagName: "div", attributes: nil)
        child.parentElement = parent
        child.textContent = "123"
        let value = self.parser!.integerContent(element: parent)!
        
        XCTAssertEqual(value, 123, "")
    }
    
    func testIntegerContent_ElementNil() {
        let value = self.parser!.integerContent(element: nil)
        
        XCTAssertNil(value, "")
    }
    
    func testIntegerContent_NoInteger() {
        let html = HTMLElement(tagName: "div", attributes: nil)
        html.textContent = "some text"
        let value = self.parser!.integerContent(element: html)
        
        XCTAssertNil(value, "")
    }
    
    func testIntegetContent_MixedContent() {
        let html = HTMLElement(tagName: "div", attributes: nil)
        html.textContent = "some text with 123 numbers 456"
        let value = self.parser!.integerContent(element: html)
        
        XCTAssertNil(value, "")
    }
}

extension ForumParserTests {
    func testPostDate_Success() {
        let html = HTMLElement(tagName: "div", attributes: nil)
        html.textContent = "10.10.2014 , 10:10 AM | #1"
        let value = self.parser!.postDate(element: html)!
        
        XCTAssertEqual(value, "10.10.2014, 10:10 AM", "")
    }
    
    func testPostDate_Nested() {
        let parent = HTMLElement(tagName: "div", attributes: nil)
        let child = HTMLElement(tagName: "div", attributes: nil)
        child.parentElement = parent
        child.textContent = "10.10.2014 , 10:10 AM | #1"
        let value = self.parser!.postDate(element: parent)!
        
        XCTAssertEqual(value, "10.10.2014, 10:10 AM", "")
    }
    
    func testPostDate_ElementNil() {
        let value = self.parser!.postDate(element: nil)
        
        XCTAssertNil(value, "")
    }
}

extension ForumParserTests {
    func testPostNumber_Success() {
        let html = HTMLElement(tagName: "div", attributes: nil)
        html.textContent = "10.10.2014 , 10:10 AM | #1"
        let value = self.parser!.postNumber(element: html, language: self.language)!
        
        XCTAssertEqual(value, 1, "")
    }
    
    func testPostNumber_Nested() {
        let parent = HTMLElement(tagName: "div", attributes: nil)
        let child = HTMLElement(tagName: "div", attributes: nil)
        child.parentElement = parent
        child.textContent = "10.10.2014 , 10:10 AM | #1"
        let value = self.parser!.postNumber(element: parent, language: self.language)!
        
        XCTAssertEqual(value, 1, "")
    }
    
    func testPostNumber_ElementNil() {
        let value = self.parser!.postNumber(element: nil, language: self.language)
        
        XCTAssertNil(value, "")
    }
}

extension ForumParserTests {
    func testPostText_Success() {
        let header = "Header text"
        let body = "Body text"
        
        let value = self.parser!.formatPostBlock(header: header, body: body)
        
        XCTAssertEqual(value, String(format: self.parser!.postBlockFormat, header, body), "")
    }
    
    func testPostText_MissingHeader() {
        let body = "Body text"
        
        let value = self.parser!.formatPostBlock(header: nil, body: body)
        
        XCTAssertEqual(value, String(format: self.parser!.postBlockNoHeaderFormat, body), "")
    }
    
    func testPostText_MissingBody() {
        let header = "Header text"
        
        let value = self.parser!.formatPostBlock(header: header, body: nil)
        
        XCTAssertEqual(value, "", "")
    }
    
    func testPostText_Simple() {
        let html = "<div>Test post text<br>More text in new line</div>"
        let doc = HTMLDocument(string: html)
        
        let value = self.parser!.postText(node: doc.rootElement)
        
        XCTAssertNotNil(value, "")
        XCTAssertEqual(value!, doc.rootElement!.textContent, "")
    }
    
    func testPostText_Nested() {
        let html = "<div><span>Test post text</span><br><span>More text in new line</span><div>Even more text</div></div>"
        let doc = HTMLDocument(string: html)
        
        let value = self.parser!.postText(node: doc.rootElement)
        
        XCTAssertNotNil(value, "")
        XCTAssertEqual(value!, doc.rootElement!.textContent, "")
    }
    
    func testPostText_Styled() {
        let html = "<div><font color='Yellow'>Yellow text</font><font color='Red'><font size='4'><b>Red bold text of size 4</b></font></font><br><font size='6'><i>Italic text of size 6</i></font></div>"
        let doc = HTMLDocument(string: html)
        
        let value = self.parser!.postText(node: doc.rootElement)
        
        XCTAssertNotNil(value, "")
        XCTAssertEqual(value!, doc.rootElement!.textContent, "")
    }
    
    func testPostText_WithBlock() {
        for blockClass in self.parser!.postBlockClasses {
            let blockHtml = "<div class='\(blockClass)'><div class='\(blockClass)-header'>Block by TestUser</div><div class='\(blockClass)-body'>Block body text here<br>More text in new line</div></div>"
            let postHtml = "<div class='regular-post'>Regular post text<br><br>More regular post text in a new paragraph</div>"
            let html = "<div>\(blockHtml)\(postHtml)</div>"
            let doc = HTMLDocument(string: html)
            
            let value = self.parser!.postText(node: doc.rootElement)
            
            let header = doc.rootElement!.firstNode(matchingSelector: ".\(blockClass)-header")!.textContent
            let body = doc.rootElement!.firstNode(matchingSelector: ".\(blockClass)-body")!.textContent
            let block = self.parser!.formatPostBlock(header: header, body: body)
            let post = doc.rootElement!.firstNode(matchingSelector: ".regular-post")!.textContent
            
            XCTAssertNotNil(value, "")
            XCTAssertEqual(value!, "\(block)\(post)", "")
        }
    }
    
    func testPostText_WithMultipleBlocks() {
        for blockClass in self.parser!.postBlockClasses {
            let block1Html = "<div class='\(blockClass)'><div class='\(blockClass)-header'>Block by TestUser</div><div class='\(blockClass)-body'>Block body text here<br>More text in new line</div></div>"
            let block2Html = "<div class='\(blockClass)'><div class='\(blockClass)-header'>Block by DifferentUser</div><div class='\(blockClass)-body'>Simple block body</div></div>"
            
            let post1Html = "<div class='regular-post'>Regular post text<br><br>More regular post text in a new paragraph</div>"
            let post2Html = "<div class='regular-post'>Simple regular post text</div>"
            
            let html = "<div>\(block1Html)\(post1Html)\(block2Html)\(post2Html)</div>"
            let doc = HTMLDocument(string: html)
            
            let value = self.parser!.postText(node: doc.rootElement)
            
            let blocks = doc.rootElement!.nodes(matchingSelector: ".\(blockClass)")
            var blocksText = Array<String>()
            for block in blocks {
                let header = block.firstNode(matchingSelector: ".\(blockClass)-header")!.textContent
                let body = block.firstNode(matchingSelector: ".\(blockClass)-body")!.textContent
                let block = self.parser!.formatPostBlock(header: header, body: body)
                blocksText.append(block)
            }
            let posts = doc.rootElement!.nodes(matchingSelector: ".regular-post")
            var postsText = Array<String>()
            for post in posts {
                postsText.append(post.textContent)
            }
            
            XCTAssertNotNil(value, "")
            XCTAssertEqual(value!, "\(blocksText[0])\(postsText[0])\(blocksText[1])\(postsText[1])", "")
        }
    }
    
    func testPostText_WithSpecialBlock() {
        let blockHtml = "<div style='margin:20px; margin-top:5px; '><div class='smallfont'>Quote:</div><table><tbody><tr><td class='alt2'><div class='header'>Block by TestUser</div><div class='body'>Block body text here<br>More text in new line</div></td></tr></tbody></table></div>"
        let postHtml = "<div class='regular-post'>Regular post text<br><br>More regular post text in a new paragraph</div>"
        
        let html = "<div>\(blockHtml)\(postHtml)</div>"
        let doc = HTMLDocument(string: html)
        
        let value = self.parser!.postText(node: doc.rootElement)
        
        let header = doc.rootElement!.firstNode(matchingSelector: ".header")!.textContent
        let body = doc.rootElement!.firstNode(matchingSelector: ".body")!.textContent
        let block = self.parser!.formatPostBlock(header: header, body: body)
        let post = doc.rootElement!.firstNode(matchingSelector: ".regular-post")!.textContent
        
        XCTAssertNotNil(value, "")
        XCTAssertEqual(value!, "\(block)\(post)", "")
    }
}

File: android/app/src/test/java/com/ifabijanovic/holonet/forum/data/ForumParserTests.kt
package com.ifabijanovic.holonet.forum.data

import org.jsoup.nodes.Element
import org.junit.Assert
import org.junit.Test

/**
 * Created by feb on 19/03/2017.
 */
class ForumParserTests {
    @Test
    fun linkParameter_Success() {
        val link = Element("a")
        link.attr("href", "http://www.holonet.test?param=value")
        val param = ForumParser().linkParameter(link, "param")

        Assert.assertNotNull(param)
        Assert.assertEquals("value", param)
    }

    @Test
    fun linkParameter_ElementNull() {
        val param = ForumParser().linkParameter(null, "param")
        Assert.assertNull(param)
    }

    @Test
    fun linkParameter_MissingParameter() {
        val link = Element("a")
        link.attr("href", "http://www.holonet.test")
        val param = ForumParser().linkParameter(link, "param")

        Assert.assertNull(param)
    }

    @Test
    fun linkParameter_MultipleParameters() {
        val link = Element("a")
        link.attr("href", "http://www.holonet.test?param1=value1&param2=value2")
        val param = ForumParser().linkParameter(link, "param2")

        Assert.assertNotNull(param)
        Assert.assertEquals("value2", param)
    }

    //

    @Test
    fun integerContent_Success() {
        val html = Element("div")
        html.text("123")
        val value = ForumParser().integerContent(html)

        Assert.assertEquals(123, value)
    }

    @Test
    fun integerContent_Nested() {
        val parent = Element("div")
        val child = Element("div")
        parent.appendChild(child)
        child.text("123")
        val value = ForumParser().integerContent(parent)

        Assert.assertEquals(123, value)
    }

    @Test
    fun integerContent_ElementNull() {
        val value = ForumParser().integerContent(null)
        Assert.assertNull(value)
    }

    @Test
    fun integerContent_NoInteger() {
        val html = Element("div")
        html.text("some text")
        val value = ForumParser().integerContent(html)

        Assert.assertNull(value)
    }

    @Test
    fun integerContent_MixedContent() {
        val html = Element("div")
        html.text("some text with 123 numbers 456 ")
        val value = ForumParser().integerContent(html)

        Assert.assertNull(value)
    }

    //

    @Test
    fun postDate_Success() {
        val html = Element("div")
        html.text("10.10.2014 , 10:10 AM | #1")
        val value = ForumParser().postDate(html)

        Assert.assertEquals("10.10.2014, 10:10 AM", value)
    }

    @Test
    fun postDate_Nested() {
        val parent = Element("div")
        val child = Element("div")
        parent.appendChild(child)
        child.text("10.10.2014 , 10:10 AM | #1")
        val value = ForumParser().postDate(parent)

        Assert.assertEquals("10.10.2014, 10:10 AM", value)
    }

    @Test
    fun postDate_ElementNull() {
        val value = ForumParser().postDate(null)
        Assert.assertNull(value)
    }

    //

    @Test
    fun postNumber_Success() {
        val html = Element("div")
        html.text("10.10.2014 , 10:10 AM | #17")
        val value = ForumParser().postNumber(html)

        Assert.assertEquals(17, value)
    }

    @Test
    fun postNumber_Nested() {
        val parent = Element("div")
        val child = Element("div")
        parent.appendChild(child)
        child.text("10.10.2014 , 10:10 AM | #77")
        val value = ForumParser().postNumber(parent)

        Assert.assertEquals(77, value)
    }

    @Test
    fun postNumber_ElementNull() {
        val value = ForumParser().postNumber(null)
        Assert.assertNull(value)
    }

    @Test
    fun postNumber_DevWithNextMultipleLanguages() {
        val en = Element("div")
        en.text("10.10.2014 , 10:10 AM | #5 Next")
        val enValue = ForumParser().postNumber(en)
        Assert.assertEquals(5, enValue)

        val fr = Element("div")
        fr.text("10.10.2014 , 10:10 AM | #77 Suivante")
        val frValue = ForumParser().postNumber(fr)
        Assert.assertEquals(77, frValue)

        val de = Element("div")
        de.text("10.10.2014 , 10:10 AM | #203 Nächste")
        val deValue = ForumParser().postNumber(de)
        Assert.assertEquals(203, deValue)
    }
}
